### PR TITLE
Create access list

### PR DESCRIPTION
**Pull request by lupin012**

### CODE REVIEW

File: silkrpc/core/evm_access_list_tracer.cpp [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+       }
+    }
+    silkworm::AccessListEntry item;
+    item.account = address;
+    access_list_.push_back(item);
+}
+
+void AccessListTracer::add_local_access_list(const std::vector<silkworm::AccessListEntry> input_access_list) {
+    for (int i = 0; i < input_access_list.size(); i++) {
+       if (!exclude(input_access_list[i].account))
```
Comment (canepat) [inline silkrpc/core/evm_access_list_tracer.cpp:131]: Add curly braces and indentation or join the two `if` conditions to improve readability

File: silkrpc/core/evm_access_list_tracer.hpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+private:
+    inline bool exclude(const evmc::address& address);
+    inline evmc::address address_from_bytes32_hex(const std::string& s);
+    void add_storage(const evmc::address& address, const evmc::bytes32& storage);
+    void add_address(const evmc::address& address);
+    void add_local_access_list(const std::vector<silkworm::AccessListEntry> input_access_list);
+
+
+    std::vector<silkworm::AccessListEntry> access_list_;
```
Comment (canepat) [inline silkrpc/core/evm_access_list_tracer.hpp:63]: Probably this data structure from Silkworm could be not the best choice for storing access list entries in this context

File: silkrpc/commands/eth_api.cpp [hunk 3]
```diff
@@ -1049,6 +1050,98 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& requ
     co_return;
 }
 
+// https://eth.wiki/json-rpc/API#eth_createAccessList
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/commands/eth_api.cpp:1053]: ok

File: silkrpc/commands/eth_api.cpp [hunk 4]
```diff
@@ -1049,6 +1050,98 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& requ
...
+    auto tx = co_await database_->begin();
+
+    try {
+        ethdb::TransactionDatabase tx_database{*tx};
+
+        const auto block_with_hash = co_await core::read_block_by_number_or_hash(*context_.block_cache, tx_database, block_number_or_hash);
+        const auto chain_id = co_await core::rawdb::read_chain_id(tx_database);
+        const auto chain_config_ptr = silkworm::lookup_chain_config(chain_id);
+
+        // MGT state_cache
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/commands/eth_api.cpp:1076]: removed

File: silkrpc/commands/eth_api.cpp [hunk 5]
```diff
@@ -1049,6 +1050,98 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& requ
...
+        StateReader state_reader{tx_database};
+        // If the gas amount is not set, extract this as it will depend on access
+        // lists and we'll need to reestimate every time
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/commands/eth_api.cpp:1079]: removed

File: silkrpc/commands/eth_api.cpp [hunk 6]
```diff
@@ -1049,6 +1050,98 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& requ
...
+
+        evmc::address to{};
+        if (call.to) {
+           to = *(call.to);
+        } else {
+           uint64_t nonce;
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/commands/eth_api.cpp:1085]: init to 0

File: silkrpc/commands/eth_api.cpp [hunk 7]
```diff
@@ -1049,6 +1050,98 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& requ
...
+              }
+              call.nonce = nonce;
+           } else {
+              nonce = *(call.nonce);
+           }
+           to = silkworm::create_address(*call.from, nonce);
+        }
+
+        std::vector<silkworm::AccessListEntry> prev;
+        std::vector<silkworm::AccessListEntry> curr;
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/commands/eth_api.cpp:1105]: ok

File: silkrpc/commands/eth_api.cpp [hunk 8]
```diff
@@ -1049,6 +1050,98 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& requ
...
+           prev = tracer->get_access_list();
+           const auto execution_result = co_await executor.call(block_with_hash.block, txn, tracer);
+           curr = tracer->get_access_list();
+           if (execution_result.pre_check_error) {
+              reply = make_json_error(request["id"], -32000, execution_result.pre_check_error.value());
+              break;
+           } else {
+              if (tracer->compare(curr, prev)) {
+                 AccessListResult access_list_result;
+                 access_list_result.access_list = tracer->get_access_list();
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/commands/eth_api.cpp:1121]: ok

File: silkrpc/commands/eth_api.cpp [hunk 9]
```diff
@@ -1049,6 +1050,98 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& requ
...
+              nonce = *(call.nonce);
+           }
+           to = silkworm::create_address(*call.from, nonce);
+        }
+
+        std::vector<silkworm::AccessListEntry> prev;
+        std::vector<silkworm::AccessListEntry> curr;
+        // check on precompiles contract is done on tracer
+
+        for(;;) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/commands/eth_api.cpp:1108]: add do /while (cond)

File: silkrpc/commands/eth_api.cpp [hunk 10]
```diff
@@ -1049,6 +1050,98 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& requ
...
+           auto tracer = std::make_shared<access_list::AccessListTracer>(call.access_list, *call.from, to);
+           EVMExecutor executor{context_, tx_database, *chain_config_ptr, workers_, block_with_hash.block.header.number};
+           silkworm::Transaction txn{call.to_transaction()};
+           prev = tracer->get_access_list();
+           const auto execution_result = co_await executor.call(block_with_hash.block, txn, tracer);
+           curr = tracer->get_access_list();
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/commands/eth_api.cpp:1114]: change organization of the loop

File: silkrpc/commands/eth_api.cpp [hunk 11]
```diff
@@ -1049,6 +1050,98 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& requ
...
+                 access_list_result.gas_used = txn.gas_limit - execution_result.gas_left;
+                 if (execution_result.error_code != evmc_status_code::EVMC_SUCCESS) {
+                    const auto error_message = EVMExecutor<>::get_error_message(execution_result.error_code, execution_result.data, false /* full_error */);
+                    access_list_result.error = error_message;
+                 }
+                 reply = make_json_content(request["id"], access_list_result);
+                 break;
+              }
+          }
+          call.set_access_list(curr);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/commands/eth_api.cpp:1131]: ok

File: silkrpc/core/evm_access_list_tracer.cpp [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+#include <evmc/instructions.h>
+#include <intx/intx.hpp>
+#include <silkworm/third_party/evmone/lib/evmone/execution_state.hpp>
+#include <silkworm/third_party/evmone/lib/evmone/instructions.hpp>
+
+#include <silkrpc/common/log.hpp>
+#include <silkrpc/common/util.hpp>
+#include <silkrpc/core/evm_executor.hpp>
+
+namespace silkrpc::access_list {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/core/evm_access_list_tracer.cpp:31]: ok I remove it

File: silkrpc/core/evm_access_list_tracer.hpp [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    AccessListTracer& operator=(const AccessListTracer&) = delete;
+
+    void set_access_list(std::vector<silkworm::AccessListEntry> ale) { access_list_ = ale; }
+    std::vector<silkworm::AccessListEntry> get_access_list() { return access_list_; }
+    bool compare(const std::vector<silkworm::AccessListEntry>& acl1, const std::vector<silkworm::AccessListEntry>& acl2);
+
+    void on_execution_start(evmc_revision rev, const evmc_message& msg, evmone::bytes_view code) noexcept override;
+    void on_instruction_start(uint32_t pc, const evmone::ExecutionState& execution_state, const silkworm::IntraBlockState& intra_block_state) noexcept override;
+    void on_execution_end(const evmc_result& result, const silkworm::IntraBlockState& intra_block_state) noexcept override {}
+    void dump(const std::string str, const std::vector<silkworm::AccessListEntry>& acl);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/core/evm_access_list_tracer.hpp:50]: can be used for debugging pourpouse

File: silkrpc/core/evm_access_list_tracer.cpp [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+
+std::string get_opcode_name(const char* const* names, std::uint8_t opcode) {
+    const auto name = names[opcode];
+    return (name != nullptr) ?name : "opcode 0x" + evmc::hex(opcode) + " not defined";
+}
+
+inline evmc::address AccessListTracer::address_from_bytes32_hex(const std::string& s) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/core/evm_access_list_tracer.cpp:38]: renames

File: silkrpc/core/evm_access_list_tracer.cpp [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+    return silkworm::to_evmc_address(bytes.value_or(silkworm::Bytes{}));
+}
+
+void AccessListTracer::on_execution_start(evmc_revision rev, const evmc_message& msg, evmone::bytes_view code) noexcept {
+    if (opcode_names_ == nullptr) {
+        opcode_names_ = evmc_get_instruction_names_table(rev);
+    }
+}
+
+void AccessListTracer::on_instruction_start(uint32_t pc, const evmone::ExecutionState& execution_state, const silkworm::IntraBlockState& intra_block_state) noexcept {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/core/evm_access_list_tracer.cpp:49]: added

File: silkrpc/core/evm_access_list_tracer.cpp [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+    assert(execution_state.msg);
+    evmc::address recipient(execution_state.msg->recipient);
+    evmc::address sender(execution_state.msg->sender);
+
+    const auto opcode = execution_state.code[pc];
+    auto opcode_name = get_opcode_name(opcode_names_, opcode);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/core/evm_access_list_tracer.cpp:55]: ok

File: silkrpc/core/evm_access_list_tracer.cpp [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+        << "   msg.gas: " << std::dec << execution_state.msg->gas
+        << "   msg.depth: " << std::dec << execution_state.msg->depth
+        << "}\n";
+
+    if (opcode_name == "SLOAD" && execution_state.stack.size() >= 1) {
+       const auto address = silkworm::bytes32_from_hex(intx::hex(execution_state.stack[0]));
+       add_storage(recipient, address);
+    } else if (opcode_name == "SSTORE" && execution_state.stack.size() >= 1) {
+       const auto address = silkworm::bytes32_from_hex(intx::hex(execution_state.stack[0]));
+       add_storage(recipient, address);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/core/evm_access_list_tracer.cpp:75]: joined

[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/core/evm_access_list_tracer.cpp:75]: ok

File: silkrpc/core/evm_access_list_tracer.cpp [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+
+
+
+inline bool AccessListTracer::exclude(const evmc::address& address) {
+       // return (address == from_ || address == to_ || is_precompiled(address)); // ADD check on precompiled when available from silkworm
+       return (address == from_ || address == to_);
+}
+
+void AccessListTracer::add_storage(const evmc::address& address, const evmc::bytes32& storage) {
+    //std::cout << "add_storage:: Address: " << address << " Storage: " << storage << "\n";
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/core/evm_access_list_tracer.cpp:100]: added

File: silkrpc/core/evm_access_list_tracer.cpp (hunk 1 above)
Comment (lupin012) [reply silkrpc/core/evm_access_list_tracer.cpp:131]: ok

File: silkrpc/core/evm_access_list_tracer.hpp (hunk 2 above)
Comment (lupin012) [reply silkrpc/core/evm_access_list_tracer.hpp:63]: yes

File: silkrpc/txpool/transaction_pool.hpp [hunk 19]
```diff
@@ -71,6 +71,26 @@ using TransactionsAwaitable = unary_awaitable<
     ::txpool::TransactionsReply
 >;
 
+using AddNonceClient = AsyncUnaryClient<
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/txpool/transaction_pool.hpp:74]: ok

File: silkrpc/txpool/transaction_pool.hpp [hunk 20]
```diff
@@ -71,6 +71,26 @@ using TransactionsAwaitable = unary_awaitable<
...
+    ::txpool::Txpool::StubInterface,
+    ::txpool::NonceRequest,
+    ::txpool::NonceReply,
+    &::txpool::Txpool::StubInterface::PrepareAsyncNonce
+>;
+
+using AddNonceAwaitable = unary_awaitable<
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/txpool/transaction_pool.hpp:81]: ok

File: silkrpc/txpool/transaction_pool.hpp [hunk 21]
```diff
@@ -154,7 +174,34 @@ class TransactionPool final {
         }
     }
 
+    asio::awaitable<Nonce_reply> nonce(const evmc::address& address) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/txpool/transaction_pool.hpp:177]: ok

File: silkrpc/txpool/transaction_pool.hpp [hunk 22]
```diff
@@ -154,7 +174,34 @@ class TransactionPool final {
...
+        const auto start_time = clock_time::now();
+        SILKRPC_DEBUG << "TransactionPool::nonce\n";
+        ::txpool::NonceRequest request{};
+        request.set_allocated_address(H160_from_address(address));
+        AddNonceAwaitable nonce_awaitable{executor_, stub_, queue_};
+        const auto reply = co_await nonce_awaitable.async_call(request, asio::use_awaitable);
+        SILKRPC_DEBUG << "TransactionPool::nonce found:" << reply.found() << " nonce: " << reply.nonce() <<
+                         " t=" << clock_time::since(start_time) << "\n";
+        co_return Nonce_reply{reply.found(), reply.nonce()};
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/txpool/transaction_pool.hpp:186]: ok

File: silkrpc/types/transaction.hpp [hunk 23]
```diff
@@ -47,6 +49,12 @@ struct TransactionWithBlock {
     Transaction transaction;
 };
 
+struct AccessListResult {
+    std::vector<silkworm::AccessListEntry> access_list;
+    std::optional<std::string> error;
+    uint64_t gas_used;
+};
+
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/types/transaction.hpp:57]: ok added in types/transation.hpp

File: silkrpc/commands/eth_api.cpp [hunk 24]
```diff
@@ -1049,6 +1050,98 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& requ
...
+        std::vector<silkworm::AccessListEntry> prev;
+        std::vector<silkworm::AccessListEntry> curr;
+        // check on precompiles contract is done on tracer
+
+        for(;;) {
+           auto tracer = std::make_shared<access_list::AccessListTracer>(call.access_list, *call.from, to);
+           EVMExecutor executor{context_, tx_database, *chain_config_ptr, workers_, block_with_hash.block.header.number};
+           silkworm::Transaction txn{call.to_transaction()};
+           prev = tracer->get_access_list();
+           const auto execution_result = co_await executor.call(block_with_hash.block, txn, tracer);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/commands/eth_api.cpp:1113]: if the tracer is allocated outside the loop after executor the program crash I think in the de-allocation of the shared-ptr

File: silkrpc/commands/eth_api.cpp [hunk 25]
```diff
@@ -1049,6 +1050,98 @@ asio::awaitable<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& requ
...
+           curr = tracer->get_access_list();
+           if (execution_result.pre_check_error) {
+              reply = make_json_error(request["id"], -32000, execution_result.pre_check_error.value());
+              break;
+           } else {
+              if (tracer->compare(curr, prev)) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/commands/eth_api.cpp:1119]: I have reorganized the loop. I prefer using explicit compare()

File: silkrpc/json/types.cpp [hunk 26]
```diff
@@ -113,6 +113,20 @@ void from_json(const nlohmann::json& json, uint256& ui256) {
 
 namespace silkworm {
 
+void from_json(const nlohmann::json& json, std::vector<silkworm::AccessListEntry>& al) {
```
Comment (canepat) [inline silkrpc/json/types.cpp:116]: Type serializers for STL containers are not necessary in `nlohmann`: it should be sufficient to define the serializer for the element type, i.e. `AccessListEntry` in this case

**Review by canepat — CHANGES_REQUESTED** [review]

**Review by canepat — APPROVED** [review]

**Pull request merged**